packages/marshal/src/rankOrder.js

```diff
@@ -0,0 +1,545 @@
+/**
+ * @typedef {[string, string]} RankCover
+ */
+
+/**
+ * @typedef {[number, number]} IndexCover
+ */
+
+/**
+ * This is the equality comparison used by JavaScript's Map and Set
+ * abstractions, where NaN is the same as NaN and -0 is the same as
+ * 0. Marshal serializes -0 as zero, so the semantics of our distributed
+ * object system does not distinguish 0 from -0.
+ *
+ * `sameValueZero` is the EcmaScript spec name for this equality comparison,
+ * but TODO we need a better name for the API.
+ *
+ * @param {any} x
+ * @param {any} y
+ * @returns {boolean}
+ */
+const sameValueZero = (x, y) => x === y || is(x, y);
+
+/**
+ * @type {[PassStyle, RankCover][]}
+ */
+const PassStyleRankAndCover = harden([
+  /* !  */ ['error', ['!', '!~']],
+  /* (  */ ['copyRecord', ['(', '(~']],
+  /* :  */ ['tagged', [':', ':~']],
+  /* ?  */ ['promise', ['?', '?~']],
+  /* [  */ ['copyArray', ['[', '[~']],
+  /* b  */ ['boolean', ['b', 'b~']],
+  /* f  */ ['number', ['f', 'f~']],
+  /* np */ ['bigint', ['n', 'p~']],
+  /* r  */ ['remotable', ['r', 'r~']],
+  /* s  */ ['string', ['s', 't']],
+  /* v  */ ['null', ['v', 'v~']],
+  /* y  */ ['symbol', ['y', 'z']],
+  /* z  */ ['undefined', ['z', '{']],
+  /* | remotable->ordinal mapping prefix: This is not used in covers but it is
+       reserved from the same set of strings. Note that the prefix is > any
+       prefix used by any cover so that ordinal mapping keys are always outside
+       the range of valid collection entry keys. */
+]);
+
+const PassStyleRank = fromEntries(
+  entries(PassStyleRankAndCover).map(([i, v]) => [v[0], Number(i)]),
+);
+setPrototypeOf(PassStyleRank, null);
+harden(PassStyleRank);
+
+/**
+ * Associate with each passStyle a RankCover that may be an overestimate,
+ * and whose results therefore need to be filtered down. For example, because
+ * there is not a smallest or biggest bigint, bound it by `NaN` (the last place
+ * number) and `''` (the empty string, which is the first place string). Thus,
+ * a range query using this range may include these values, which would then
+ * need to be filtered out.
+ *
+ * @param {PassStyle} passStyle
+ * @returns {RankCover}
+ */
+export const getPassStyleCover = passStyle =>
+  PassStyleRankAndCover[PassStyleRank[passStyle]][1];
+harden(getPassStyleCover);
+
+/**
+ * @type {WeakMap<RankCompare,WeakSet<Passable[]>>}
+ */
+const memoOfSorted = new WeakMap();
+
+/**
+ * @type {WeakMap<RankCompare,RankCompare>}
+ */
+const comparatorMirrorImages = new WeakMap();
+
+export const recordParts = record => {
+  assertRecord(record);
+  // TODO Measure which is faster: a reverse sort by sorting and
+  // reversing, or by sorting with an inverse comparison function.
+  // If it makes a significant difference, use the faster one.
+  const names = ownKeys(record)
+    .sort()
+    .reverse();
+  // @ts-expect-error It thinks name might be a symbol, which it doesn't like.
+  const vals = names.map(name => record[name]);
+  return harden([names, vals]);
```

Comment:
I think we can also avoid some unnecessary traversal.
```suggestion
  return freeze([names, vals]);
```

Reply:
I prefer that we stay in the robust `harden` habit and only do otherwise when there's a good enough reason. Performance can be a good enough reason, but only with good enough evidence that it makes a significant overall difference.

Leaving as is. Resolving.